docs/articles/improve-raptor-with-rag.md

```diff
@@ -0,0 +1,561 @@
+3. **UMAP**: Uniform Manifold Approximation and Projection (UMAP) is a powerful dimensionality reduction algorithm we’ll use to shrink the size of our data point embeddings. This reduction makes it easier for clustering algorithms like GMM to cluster high-dimensional embeddings.
+4. **BIC and Elbow Method**: Both techniques help identify the optimal number of clusters in a dataset. The Bayesian Information Criterion (BIC) evaluates models based on their fit to the data while penalizing complexity. The Elbow Method plots explained variance against the number of clusters, helping to pinpoint where adding more clusters offers diminishing returns. For our purposes, we’ll leverage both methods to determine the best number of clusters.
+
+### How it actually works?
+
+Now that you’re familiar with the key terms (and if not, no worries—you’ll catch on as we go!), let’s dive into how everything actually works under the hood of RAPTOR. 
+
+- **Starting Documents (Leaves)**: RAPTOR first splits raw documents into chunks.
+- **Embedding and Clustering**: Next, we embed these chunks ([RAPTOR authors](https://arxiv.org/pdf/2401.18059) use the UMAP dimensionality reduction algorithm to minimize chunk embedding size). Then, based on their vector embeddings, we cluster our text chunks (using Gaussian Mixture Models (GMM) to ensure effective grouping of high-dimensional vector embeddings).
+- **Summarizing Clusters**: Next, we summarize the clusters of similar chunks into a node of higher-level abstractions. 
+- **Recursive Iteration**: We treat these nodes in turn as chunks, clustering them, summarizing these clusters.. iteratively, building a tree-like structure that encapsulates the relationships and hierarchical structures inherent in the raw documents we started from.
+
+![image.png](../assets/use_cases/improve-raptor-with-rag/raptor-2.png)
+
+### Building the RAPTOR tree
+
+Now that we’ve unpacked how it all works (and you’re still with me hopefully, right?), let’s shift gears and talk about how we actually build the RAPTOR tree. 
+
+**Setup and Imports**
+
+```python
+pip install lancedb scikit-learn openai torch sentence_transformers tiktoken umap-learn PyPDF2
+```
+
+```python
+import os
+import uuid
+import tiktoken
+import re
+import numpy as np
+import pandas as pd
+import transformers
+import torch
+import umap.umap_ as umap
+import matplotlib.pyplot as plt
+from openai import OpenAI
+from typing import List, Tuple, Optional, Dict
+from sklearn.mixture import GaussianMixture
+from sentence_transformers import SentenceTransformer
+
+openai_api_key = "sk-XXXXXXXXXXXXXXX"
+client = OpenAI(api_key=openai_api_key)
+```
+
+### Creating the Chunks
+
+Setting up RAPTOR is pretty straightforward and builds on what we’ve already covered. The first step is to break down our textual documents into smaller chunks. Once we have those, we can convert them into dense vector embeddings.
+
+```python
+import os
```

Comment:
we need to import os twice?

Reply:
We can remove this if the `os` is already imported in previous code snippets